### PR TITLE
Examples/tp/hello_tp.py is modified to use temporal_memory.py instead Fixes #2548

### LINKED ISSUE #2548: Change hello_tp.py to use temporal_memory.py
The file `examples/tp/hello_tp.py` is for newbies who want to learn about temporal memory. It currently uses `TP10X2`. This issue is to port the file over to use the newer and easier to understand TemporalMemory.  The language in the file should also be updated to use the term "Temporal Memory" as it still refers to the older terminology "temporal pooler".  

Specifically:

* Change name of file to hello_tm.py
* Change import to `from nupic.research.temporal_memory import TemporalMemory`
* TemporalMemory has a slightly different interface than TP10X2 so the code needs to be modified accordingly.
* TemporalMemory does not have helper functions to print the cell states, so those should be implemented.
* Change terminology in comments, etc. to say "temporal memory" and "TM" instead of "temporal pooler" and "TP"

### PR DESCRIPTION
**Pull request by arhik**
Hello All,

I have modified the example/tp/hello_tp.py to use temporal_memory instead of TP10X2.py..

The comments are changed aswell according to my understanding. Rectify me if i am wrong anywhere.
This issue was raised by Subutai and partly Fixes #2548. I should get familiar with git from contribution perspective i guess. Guide me if i am not doing it right way.

Karthik

### CODE REVIEW

**Comment by arhik** [comment]
Oops i didn't change the name of file to hello_tm.py. Should i pull a new request?

**Comment by breznak** [comment]
@arhik nice work, thanks! :+1: 

> Oops i didn't change the name of file to hello_tm.py. Should i pull a new request?

you can do `git mv ...` and commit this change and push to your branch, it will appear in this PR.

File: examples/tp/hello_tp.py [hunk 1]
```diff
@@ -6,15 +6,15 @@
 # following terms and conditions apply:
 #
 # This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU Affero Public License version 3 as
```
Comment (breznak) [inline examples/tp/hello_tp.py]: correct me if I'm wrong, but I think we've switched to AGPL from GPL, so the license lines should stay.

File: examples/tp/hello_tp.py [hunk 2]
```diff
@@ -39,13 +39,19 @@
...
 import numpy
 
+# izip for maximum efficiency
+from itertools import izip as zip, count
+
 # here we choose TP implementation: (uncomment one only)
-# This is the class corresponding to the C++ optimized Temporal Pooler (default)
-from nupic.research.TP10X2 import TP10X2 as TP
+# This is the class corresponding to the C++ optimized Temporal Pooler(default)
+# from nupic.research.TP10X2 import TP10X2 as TP
```
Comment (breznak) [inline examples/tp/hello_tp.py]: does this code still run with TP? it would be nice if it did, if not, you can remove the comments here.

File: examples/tp/hello_tp.py [hunk 3]
```diff
@@ -39,13 +39,19 @@
...
 # This is simple implementation in Python
-#from nupic.research.TP import TP as TP
+
+from nupic.research.temporal_memory import TemporalMemory as TM
```
Comment (breznak) [inline examples/tp/hello_tp.py]: if there's already c++ bindings for TM (?) you could show how to import them here as well, like it was done with TP.

File: examples/tp/hello_tp.py (hunk 1 above)
Comment (arhik) [inline examples/tp/hello_tp.py]: sorry about that..

**Comment by arhik** [comment]
@ breznak 

Thank you. I will fix them soon.

File: examples/tp/hello_tp.py (hunk 3 above)
Comment (arhik) [inline examples/tp/hello_tp.py]: I am referring to FastTemporalMemory  and use of C++ Connections data structure for optimization. I hope that helps.

**Comment by rhyolight** [comment]
@arhik If you have not already, please fill out our [contributor license](http://numenta.org/licenses/cl/). Thanks!

**Comment by arhik** [comment]
I filled contributor license form. I received a confirmation email too.

**Comment by subutai** [comment]
@arhik Very nice, thanks!  Once the license text and contributor checks are passing I am :+1:

File: examples/tp/hello_tm.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 # following terms and conditions apply:
 #
 # This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU Affero Public License version 3 as
+# it under the terms of the GNU General Public License version 3 as
```
Comment (rcrowder) [inline examples/tp/hello_tm.py]: An intentional change back to GPL?

Comment (rhyolight) [inline examples/tp/hello_tm.py]: Good catch Richard. @arhik Please be sure to revert these changes from GPL to AGPL.

**Comment by rhyolight** [comment]
test

**Comment by rhyolight** [comment]
@arhik As soon as you have changed the licenses back to Affero General Public License, this is okay to merge.

File: examples/tp/hello_tm.py (hunk 4 above)
Comment (arhik) [inline examples/tp/hello_tm.py]: @rcrowder 
@rhyolight 
Sorry. My bad. That was not intentional. I worked on old code with old license header in the initial commit (didn't realize untill it was pointed out).  While trying to make recommended changes, I changed only one reference of license. I did not observe other line of it. I will be careful from now on.
